client/Windows/wf_defaults.c

```diff
@@ -0,0 +1,89 @@
+#include <Windows.h>
+#include <wincred.h>
+#include <stdio.h>
+#include <malloc.h>
+#include <freerdp/settings.h>
+
+static PCWSTR ValidateString(const BYTE* pb, ULONG cb)
+{
+	if (!pb || !cb)
+		return 0;
+
+	if (cb & (sizeof(WCHAR) - 1))
+		return 0;
+
+	if (*(WCHAR*)(pb + cb - sizeof(WCHAR)))
+		return 0;
+
+	return (PCWSTR)pb;
+}
+
+static void AddDefaultSettings_I(rdpSettings* settings, size_t idHostname, size_t idUsername,
+                          size_t idPassword)
+{
+	PCSTR ServerHostname = freerdp_settings_get_string(settings, idHostname);
+
+	if (!ServerHostname)
+		return;
+
+	BOOL bExistUserName = freerdp_settings_get_string(settings, idUsername) != 0;
+	BOOL bExistPassword = freerdp_settings_get_string(settings, idPassword) != 0;
+
+	if (bExistUserName && bExistPassword)
+		return;
+
+	PCREDENTIALA Credential;
+	PCWSTR lpWideCharStr;
+
+	PSTR Password;
+	PSTR UserName;
+
+	PSTR TargetName = 0;
+	int len = _vsnprintf(TargetName, 0, "TERMSRV/%s", (va_list)&ServerHostname);
```

Comment:
this is wrong, use `_snprintf`  (or what is current on windows) if you want variable arguments or create a proper `va_list`

Reply:
@akallabeth can you help me out here? when I change it to _snprintf, it will not work. The resulting string is completely messed up:

```C
	PSTR TargetName = 0;
	int len = _snprintf(TargetName, 0, "TERMSRV/%s", &ServerHostname);
	len++;
	TargetName = (PSTR)malloc(len);

	if (!TargetName)
		return;

	_snprintf(TargetName, len, "TERMSRV/%s", &ServerHostname);
```

TargetName is then `TERMSRV/ðŽEÎ`

Reply:
@StefanKoell you must not specify the address of the string (`&`) but the string itself in the argument list:

`int len = _snprintf(TargetName, 0, "TERMSRV/%s", ServerHostname);`

Reply:
Just tried that but I don't get a valid/correct string in this case. I'm surely missing something.

Reply:
Please disregard my last comment. It is working and I'm an idiot ;)